src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Unix.cs

```diff
@@ -171,35 +171,27 @@ internal static unsafe void WriteGatherAtOffset(SafeFileHandle handle, IReadOnly
                 stackalloc Interop.Sys.IOVector[IovStackThreshold] :
                 new Interop.Sys.IOVector[buffersCount];
 
-            try
+            long totalBytesToWrite = 0;
+            for (int i = 0; i < buffersCount; i++)
             {
-                int buffersOffset = 0, firstBufferOffset = 0;
-                while (true)
-                {
-                    long totalBytesToWrite = 0;
-
-                    for (int i = buffersOffset; i < buffersCount; i++)
-                    {
-                        ReadOnlyMemory<byte> buffer = buffers[i];
-                        totalBytesToWrite += buffer.Length;
-
-                        MemoryHandle memoryHandle = buffer.Pin();
```

Comment:
So far, for incomplete writes we were pinning the memory twice. I am not sure if this can create some kind of edge case bugs, but I think we can do it just once, before we enter the main loop